File: medium/cpp/2743.cpp
// https://leetcode.com/problems/count-substrings-without-repeating-character/description/

class Solution
{
public:
    int numberOfSpecialSubstrings(std::string s)
    {
        const auto n = std::size(s);

        auto res = 0;

        std::array<int, 26> cnt { 0 };

        for (auto left = 0, right = 0; right < n; right++)
        {
            cnt[s[right] - 'a']++;

            while (cnt[s[right] - 'a'] == 2)
            {
                cnt[s[left++] - 'a']--;
            }

            res += right - left + 1;
        }

        return res;
    }
};

File: medium/python/2743.py
# https://leetcode.com/problems/count-substrings-without-repeating-character/description/

class Solution:
    def numberOfSpecialSubstrings(self, s: str) -> int:
        freq = defaultdict(int)
        l = 0
        res = 0

        for r in range(0, len(s)):
            freq[s[r]] += 1

            while freq[s[r]] > 1:
                freq[s[l]] -= 1
                l += 1

            res += r - l + 1

        return res